src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -100,44 +106,78 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     	if (cacheService.isContainComponent()) { // Item is in cache
     		cacheItem = cacheService.getCacheOfComponent();
     		if (cacheItem.isExpired()) { // cacheItem has expired
-    			// Update the cache in a separate thread
-    			populateCacheTask(cacheItem);
+    			// Refresh the cacheItem in a separate thread
+    			refreshCacheItemTask(cacheItem);
     		}
-    	} else { // Item is not in cache
-    		// Create a new cacheItem object to be stored in cache
-    		cacheItem = new MessageCacheItem();
-    		fetchMessages(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().iterator());
-
-			if (!cacheItem.getCachedData().isEmpty()) {
-				cacheService.addCacheOfComponent(cacheItem);
-			} else if (!dto.getLocale().equals(ConstantsKeys.SOURCE) && fallbackLocalesIter!=null && fallbackLocalesIter.hasNext()) {
-    			// If failed to fetch message, use MessageCacheItem of the next fallback locale.
-				MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), fallbackLocalesIter.next().toLanguageTag(), dto.getProductID(), dto.getVersion());
-				cacheItem = new ComponentService(fallbackLocaleDTO).getMessages(fallbackLocalesIter);
-				if (!cacheItem.getCachedData().isEmpty()) {
-					cacheService.addCacheOfComponent(cacheItem);
-				}
+			// If the cacheItem is for a fallback locale, create and store cacheItem for the requested locale in a separate thread.
+    		if (!LocaleUtility.isSameLocale(cacheItem.getLocale(), this.dto.getLocale())) {
+				this.createCacheItemTask(fallbackLocalesIter);
 			}
```

Comment:
We actually want to always try to fetch every time in a separate thread, not just when the cacheItem has expired. This is because if the cacheItem is for a fallback locale, many other calls for non-supported locales can trigger a refresh  of the fallback locale cacheItem. In that case, it may never enter this if condition if we put it inside the above if block,